### PR TITLE
Add Covariates to `ParetoNBDModel`

### PR DESCRIPTION
**Pull request by ColtAllen**
## Description

This PR provides the option of using time-invariant covariates with `ParetoNBDModel`. This is a draft PR for now because I still need to parametrize some tests and add a few other things, but more importantly get opinions on two points:

Covariates require scaling for model convergence, preferably with `MinMaxScaler` so the coefficients are interpretable. This can easily be added as a utility function, and the covariate Min/Max values saved as `model.idata` attributes for out-of-sample predictions. **Should a scaler be added in this PR, or a future one? Should it be added to `ParetoNBDModel` , or the base `CLVModel`?**

Covariates for existing customers will also require modifications to the "new customer" methods. There are three options for this:

1. Just raise a `ValueError` indicating these methods aren't supported with covariates (easy, but not my preference.)
2. Average all the customer-specific `alpha` and `beta` parameters together for new customer estimates, which is a subset of the next option. 
3. Add arguments to these methods for user-specified covariate values. This will be involved and maybe even clunky to use in practice, but if a user wants to do sensitivity analysis on how covariates can impact new customer estimates, this is the way to go.

**Regardless of the approach, should it added to this PR, or another one? This one is already rather lengthy.**

## Related Issue

- [ ] Closes #
- [x] Related to # https://github.com/pymc-labs/pymc-marketing/issues/183, https://github.com/pymc-labs/pymc-marketing/issues/134

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [ ] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--463.org.readthedocs.build/en/463/

### CODE REVIEW

**Comment by ColtAllen** [comment]
Tests are all passing, albeit incomplete. The CI seems to have timed out on an MMM test.

**Comment by twiecki** [comment]
Ready for review?

**Comment by ColtAllen** [comment]
> Ready for review?

I'm gonna saves the covariate scaling and new customer modifications for future PRs. After all the tests are parametrized for covariates this will be ready for review.

**Comment by ColtAllen** [comment]
@ricardoV94 I could use your help with an `xarray` dimensional error in the `expected_purchases` and `expected_probability_alive` methods:

```
 ValueError: cannot reindex or align along dimension 'customer_id' 
because of conflicting dimension sizes: {2356, 2357} 
(note: an index is found along that dimension with size=2356)
```

Otherwise, this is ready for review.

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 1]
```diff
@@ -144,63 +152,107 @@ class ParetoNBDModel(CLVModel):
...
+            raise ValueError("Customers must have unique ID labels.")
+
+        # validate data contains specified column names for covariates
+        for _ in zip(
+            [pr_covar_columns, dr_covar_columns],
+            ["purchase_rate_covariates", "dropout_covariates"],
+        ):
+            if _[0] is not None:
+                self._validate_column_names(data, _[0])
+                # self.coords[_[1]] = _[0]  # type: ignore
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/pareto_nbd.py:194]: shall we remove this commented code?

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 2]
```diff
@@ -219,31 +281,117 @@ def build_model(  # type: ignore
...
+        return r, alpha, s, beta
+
+    def _process_covariates(
+        self,
+        data: Union[pd.DataFrame, None],
+    ) -> Tuple[Any, Any]:
+        """
+        Utility function to process covariates into model parameters.
+        """
+        # TODO: broadcasting needs to be resolved for out-of-sample data
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/pareto_nbd.py:362]: Do you want to add a Warning to handle this TODO in case is not part of the PR?

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 3]
```diff
@@ -410,15 +559,17 @@ def expected_purchases_new_customer(
...
+        # TODO: If requested, this is still possible from a what-if standpoint.
+        if self.pr_covar_columns or self.pr_covar_columns is not None:
+            raise NotImplementedError(
+                "New customer estimates not currently supported with covariates"
+            )
 
         t = np.asarray(t)
 
-        r, alpha, s, beta = self._unload_params()
+        # TODO: _unload_params requires modification for covariates w/ new customers
```
Comment (juanitorduz) [inline pymc_marketing/clv/models/pareto_nbd.py:571]: is this TODO part of the PR?

**Comment by juanitorduz** [comment]
@ColtAllen This is very cool! I need (and will!) to catch up with the development! I left some basic comments on the PR. I hope I can contribute more once I read the model details 🙏

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 4]
```diff
@@ -70,12 +69,21 @@ class ParetoNBDModel(CLVModel):
             * `recency`: time between the first and the last purchase
             * `T`: time between the first purchase and the end of the observation period; model assumptions require T >= recency
             * `customer_id`: unique customer identifier
+        Along with optional covariate columns.
+    pr_covar_columns: list, optional
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:73]: More readable name?
```suggestion
    purchase_covariate_cols: list, optional
```

Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:73]: More readable name?
```suggestion
    purchase_covariate_cols: list, optional
```

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 5]
```diff
@@ -70,12 +69,21 @@ class ParetoNBDModel(CLVModel):
...
+        List containing column names of covariates for customer purchase rates.
+    dr_covar_columns: list, optional
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:75]: More readable name?
```suggestion
    dropout_covariate_cols: list, optional
```

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 6]
```diff
@@ -70,12 +69,21 @@ class ParetoNBDModel(CLVModel):
...
-            * `beta_prior`: Scale parameter of time until churn for customer population; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `r_prior`: Shape parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `alpha_prior`: Scale parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `s_prior`: Shape parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `beta_prior`: Scale parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `alpha0_prior: Scale parameter of time between purchases if using covariates; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `beta0_prior: Scale parameter of time until dropout if using covariates; ; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `pr_prior`: Coefficients for purchase rate covariates; defaults to `pymc.Normal.dist(mu=0, sigma=1, n=len(tr_covar)`
+            * `dr_prior`: Coefficients for dropout covariates; defaults to `pymc.Normal.dist(mu=0, sigma=1, n=len(dr_covar)`
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:86]: If you agree to change the names of the cols, I would call those the same, just replace `cols` by `prior`

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 7]
```diff
@@ -144,63 +152,107 @@ class ParetoNBDModel(CLVModel):
...
-        except KeyError:
-            raise KeyError("T column is missing from data")
+        # TODO: Write a separate, more comprehensive RFM validation method in a future PR.
+        # Perform validation checks on column names and customer identifiers
+        self._validate_column_names(data, ["customer_id", "frequency", "recency", "T"])
+        if len(np.unique(data["customer_id"])) != len(data["customer_id"]):
+            raise ValueError("Customers must have unique ID labels.")
+
+        # validate data contains specified column names for covariates
+        for _ in zip(
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:188]: Give a readable name to the `_` variable. Seems like zip is no longer needed either?

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 8]
```diff
@@ -70,12 +69,21 @@ class ParetoNBDModel(CLVModel):
...
-            * `r_prior`: Shape parameter of time between purchases for customer population; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
-            * `alpha_prior`: Scale parameter of time between purchases for customer population; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
-            * `s_prior`: Shape parameter of time until churn for customer population; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
-            * `beta_prior`: Scale parameter of time until churn for customer population; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `r_prior`: Shape parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `alpha_prior`: Scale parameter of time between purchases; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `s_prior`: Shape parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=1)`
+            * `beta_prior`: Scale parameter of time until dropout; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
+            * `alpha0_prior: Scale parameter of time between purchases if using covariates; defaults to `pymc.Weibull.dist(alpha=2, beta=10)`
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:83]: Why not reuse the kwarg `alpha_prior` since only one of them is used anyway depending on covariates?

File: pymc_marketing/clv/models/pareto_nbd.py [hunk 9]
```diff
@@ -219,31 +281,117 @@ def build_model(  # type: ignore
         with pm.Model(coords=self.coords) as self.model:
             # purchase rate priors
             r = self.model.register_rv(self.r_prior, name="r")
-            alpha = self.model.register_rv(self.alpha_prior, name="alpha")
+            if self.pr_covar_columns is not None:
+                alpha0 = self.model.register_rv(self.alpha0_prior, name="alpha0")
```
Comment (ricardoV94) [inline pymc_marketing/clv/models/pareto_nbd.py:285]: instead of `alpha0` I would call this `alpha_scale`. The zero doesn't convey any meaning. Same for `beta0`

File: tests/clv/models/test_pareto_nbd.py [hunk 10]
```diff
@@ -265,6 +356,8 @@ def test_expected_probability_alive(self):
         est_prob_alive_t = self.model.expected_probability_alive(future_t=4.5)
         assert est_prob_alive.mean() > est_prob_alive_t.mean()
 
+        self.covar_model.expected_probability_alive()
```
Comment (ricardoV94) [inline tests/clv/models/test_pareto_nbd.py:359]: Test something about this. Doesn't the same question of covariates show up here?

File: tests/clv/models/test_pareto_nbd.py [hunk 11]
```diff
@@ -290,6 +383,8 @@ def test_expected_purchase_probability(self, test_n, test_t):
             rtol=0.001,
         )
 
+        self.model.expected_purchase_probability(test_n, test_t, self.data)
```
Comment (ricardoV94) [inline tests/clv/models/test_pareto_nbd.py:386]: Did you mean `self.covar_model`? Also need to test something about the output

File: tests/clv/models/test_pareto_nbd.py [hunk 12]
```diff
@@ -338,8 +433,8 @@ def test_posterior_distributions(self, fake_fit, T) -> None:
         assert customer_purchase_rate.shape == expected_shape
         assert customer_rec_freq.shape == expected_pop_dims
 
-        lam = pm.Gamma.dist(alpha=self.r_true, beta=1 / self.alpha_true, size=N)
-        mu = pm.Gamma.dist(alpha=self.s_true, beta=1 / self.beta_true, size=N)
+        lam = pm.Gamma.dist(alpha=self.r_true, beta=self.alpha_true, size=N)
+        mu = pm.Gamma.dist(alpha=self.s_true, beta=self.beta_true, size=N)
```
Comment (ricardoV94) [inline tests/clv/models/test_pareto_nbd.py:437]: Why this change?

**Comment by ricardoV94** [comment]
> * Just raise a `ValueError` indicating these methods aren't supported with covariates (easy, but not my preference.)
> 
> * Average all the customer-specific `alpha` and `beta` parameters together for new customer estimates, which is a subset of the next option.
> 
> * Add arguments to these methods for user-specified covariate values. This will be involved and maybe even clunky to use in practice, but if a user wants to do sensitivity analysis on how covariates can impact new customer estimates, this is the way to go.

I don't like option 2. I would say 1 if you want to get this merged quickly, and 3 otherwise (or later)

**Comment by ricardoV94** [comment]
> @ricardoV94 I could use your help with an `xarray` dimensional error in the `expected_purchases` and `expected_probability_alive` methods:
> 
> ```
>  ValueError: cannot reindex or align along dimension 'customer_id' 
> because of conflicting dimension sizes: {2356, 2357} 
> (note: an index is found along that dimension with size=2356)
> ```
> 
> Otherwise, this is ready for review.

What code can I look at that runs this error? The dev notebook/ a specific test?

File: pymc_marketing/clv/models/pareto_nbd.py (hunk 3 above)
Comment (ColtAllen) [reply pymc_marketing/clv/models/pareto_nbd.py:571]: It is not. Supporting new customer estimates with covariates will be involved (more details in my original message.) I'll hold off on adding this until there is a demand for it.

File: tests/clv/models/test_pareto_nbd.py (hunk 12 above)
Comment (ColtAllen) [reply tests/clv/models/test_pareto_nbd.py:455]: The beta parameters weren't specified properly. More details here:
https://github.com/pymc-labs/pymc-marketing/issues/328

**Comment by ColtAllen** [comment]
> > ```
> >  ValueError: cannot reindex or align along dimension 'customer_id' 
> > because of conflicting dimension sizes: {2356, 2357} 
> > (note: an index is found along that dimension with size=2356)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   

> 
> What code can I look at that runs this error? The dev notebook/ a specific test?

Tests are failing at these lines:
https://github.com/ColtAllen/pymc-marketing/blob/pareto_covar/pymc_marketing/clv/models/pareto_nbd.py#L687

https://github.com/ColtAllen/pymc-marketing/blob/pareto_covar/pymc_marketing/clv/models/pareto_nbd.py#L639

https://github.com/ColtAllen/pymc-marketing/blob/pareto_covar/pymc_marketing/clv/models/pareto_nbd.py#L554

File: tests/clv/models/test_pareto_nbd.py (hunk 11 above)
Comment (ColtAllen) [reply tests/clv/models/test_pareto_nbd.py]: `xarray` issue needs to be resolved first (see my latest comment).

File: tests/clv/models/test_pareto_nbd.py (hunk 10 above)
Comment (ColtAllen) [reply tests/clv/models/test_pareto_nbd.py]: `xarray` issue needs to be resolved first (see my latest comment).

**Comment by ColtAllen** [comment]
I don't know how important this is to resolve, but PyCharm is highlighting this inconsistency:

```
Signature of method 'ParetoNBDModel.build_model()' does not match signature of the base method in class 'ModelBuilder' 
 Inspection info:
Reports inconsistencies in overriding method signatures.
Example:
class Book:
    def add_title(self):
        pass


class Novel(Book):
    def add_title(self, text):
        pass
Parameters of the add_title method in the Novel class do not match the method signature specified in the Book class. As a fix, the IDE offers to apply the Change Signature refactoring.
```
